### PR TITLE
[P4Testgen] Add a DPDK-PTF P4Testgen back end and the corresponding test runner

### PR DESCRIPTION
**Pull request by Hoooao**
Improved based on #4072 to use Testgen to generate PTF tests for pna/dpdk p4 programs.

### CODE REVIEW

File: .github/workflows/PTF-TEST.yml [hunk 1]
```diff
@@ -0,0 +1,100 @@
+name: "PTF-backend"
+# This is just a temporary workflow to test the PTF backend
+# Will be merged to ci-dpdk-ptf-tests.yml once it is ready
+on:
+  push:
+    branches:
+      - ptf-test-pr
```
Comment (fruffy) [inline .github/workflows/PTF-TEST.yml:7]: This test needs to be updated. I would recommend replacing the existing Github action workflow.

File: testdata/p4_16_samples/pna-dpdk-small_sample.p4 [hunk 2]
```diff
@@ -0,0 +1,180 @@
+/*
```
Comment (fruffy) [inline testdata/p4_16_samples/pna-dpdk-small_sample.p4:1]: This example does not have a corresponding reference file. They need to be generated.

File: .github/workflows/PTF-TEST.yml (hunk 1 above)
Comment (Hoooao) [reply .github/workflows/PTF-TEST.yml:8]: yes, it will replace the existing dpdk ptf ci.

**Comment by jafingerhut** [comment]
@Hoooao @fruffy Question about the goal of this PR: When I compile the latest version of p4testgen, install it, and then try running a command like this:
```
$ cd p4c/testdata/p4_16_samples
$ p4testgen --target dpdk --arch pna --max-tests 1000 --out-dir tmp --test-backend ptf pna-dpdk-header-stack-assignment.p4 
In file: /home/andy/forks/p4c/backends/p4tools/modules/testgen/targets/pna/test_backend.cpp:49
Unimplemented compiler support: Test back end PTF not implemented for this target. Supported back ends are [METADATA].
```
Is the purpose of this PR to somehow use p4testgen to auto-generate PTF tests and execute them?  What p4testgen command line do you use to accomplish the generation of the PTF tests?

**Comment by fruffy** [comment]
> @Hoooao @fruffy Question about the goal of this PR: When I compile the latest version of p4testgen, install it, and then try running a command like this:
> 
> ```
> $ cd p4c/testdata/p4_16_samples
> $ p4testgen --target dpdk --arch pna --max-tests 1000 --out-dir tmp --test-backend ptf pna-dpdk-header-stack-assignment.p4 
> In file: /home/andy/forks/p4c/backends/p4tools/modules/testgen/targets/pna/test_backend.cpp:49
> Unimplemented compiler support: Test back end PTF not implemented for this target. Supported back ends are [METADATA].
> ```
> 
> Is the purpose of this PR to somehow use p4testgen to auto-generate PTF tests and execute them? What p4testgen command line do you use to accomplish the generation of the PTF tests?

Yes, the goal is be able to generate PTF tests for the DPDK PNA architecture using the command you described.

**Comment by jafingerhut** [comment]
@Hoooao It is obviously up to you when you want to consider this feedback relative to other changes you are making on this PR, but note that there are currently unresolved conflicts in a couple of files between your changes, and some files that have had commits merged in that change them.  You can see the "This branch has conflicts that must be resolved" messages on the Github PR page for details.

**Comment by Hoooao** [comment]
> @Hoooao It is obviously up to you when you want to consider this feedback relative to other changes you are making on this PR, but note that there are currently unresolved conflicts in a couple of files between your changes, and some files that have had commits merged in that change them. You can see the "This branch has conflicts that must be resolved" messages on the Github PR page for details.

Thanks for the heads up

**Comment by jafingerhut** [comment]
@Hoooao Do you consider this read for review now, and then merging if reviews are OK?  Or you expect to make more commits before review?

**Comment by jafingerhut** [comment]
FYI, in case it might be related to the reason you are seeing P4-DPDK fail to start, if you have a number of ports that is not a power of 2, it fails when you try to load a P4 program into it (not when it starts up, but when you try to do the SetForwardingPipelineConfig to load the P4 program).

If you are having issues where infrap4d fails completely to start up at all, not even to the point that it should be waiting for a SetForwardingPipelineConfig request, then I do not have any guesses why that may be.

**Comment by fruffy** [comment]
> If you are having issues where infrap4d fails completely to start up at all, not even to the point that it should be waiting for a SetForwardingPipelineConfig request, then I do not have any guesses why that may be.

It appears to be crashing when trying to launch `bf_switchd`. Unclear why unfortunately.

**Comment by jafingerhut** [comment]
My belief is that there should be no reason to run both infrap4d and bf_switchd.

bf_switchd gives TDI access to table add/modify/delete operations in the P4-DPDK data plane.

infrap4d contains a P4Runtime API server based on the Stratum code that listens for incoming TCP connection requests from P4Runtime API clients, and translates request messages into the appropriate TDI add/modify/delete operations.

Running both at the same time would likely only lead to problems, would be my guess.

**Comment by jafingerhut** [comment]
> `infrap4d` launches `bf_switchd` [here](https://github.com/ipdk-io/stratum-dev/blob/339022ac39df6678261ed13ecb4c35528087d1ed/stratum/hal/lib/tdi/dpdk/dpdk_sde_target.cc#L82). Would love to avoid calling it but that seems to be baked in.

I was unaware of that.  Thanks for the info.

My statement is still correct, I think, from the perspective that there is no reason if you are manually running commands in sequence, to start both infrap4d and bf_switchd _from the command line_.  You pick one.

**Comment by fruffy** [comment]
> My statement is still correct, I think, from the perspective that there is no reason if you are manually running commands in sequence, to start both infrap4d and bf_switchd _from the command line_. You pick one.

In our case we are calling `infrap4d`, which in turns calls into the `bf_switchd` library, which then crashes. We can only reproduce this in CI, but it is unfortunately a show stopper to develop a functional testing pipeline for the PTF DPDK target.

**Comment by fruffy** [comment]
@jafingerhut In this run https://github.com/p4lang/p4c/actions/runs/6501241470/job/17658197303?pr=4173 infrap4d fails to start (and appears to crash). This seemingly only happens in CI. Do you know anyone we could poke about this?

**Comment by jafingerhut** [comment]
@bharticemk @5abeel @satish153 @swaroopsarma Would one of you know why @fruffy (Fabian) might be experiencing the problems described in this comment: https://github.com/p4lang/p4c/pull/4173#issuecomment-1761700086 ?

I have run infrap4d within an IPDK Networking docker container before, but I have not yet attempted to do so outside of a Docker container, which I believe Fabian is trying to do here.  I do not have a good summary of exactly how he has built and runs these tests, but I believe all of the details are available in this PR and the code it is based upon.

File: backends/dpdk/base_test.py [hunk 3]
```diff
@@ -0,0 +1,1142 @@
+# This file was copied and then modified from the file
```
Comment (fruffy) [inline backends/dpdk/base_test.py:1]: There is a lot of redundancy here. We can likely make this a shared library for both BMv2 and DPDK.

File: backends/dpdk/run-dpdk-ptf-test.py [hunk 4]
```diff
@@ -233,19 +248,21 @@ def build_and_load_pipeline(
...
-            f"{self.options.ipdk_install_dir}/bin/p4rt-ctl "
-            "set-pipe br0 "
-            f"{conf_bin} "
-            f"{info_name} "
-        )
-        returncode = self.bridge.ns_exec(command, timeout=30)
-        if returncode != testutils.SUCCESS:
-            testutils.log.error("Failed to load pipeline")
-            return returncode
+        # NOTE: in generated PTF tests, the pipelines are loaded in individual test cases.
```
Comment (fruffy) [inline backends/dpdk/run-dpdk-ptf-test.py:251]: This comment should be removed. Fewer moving parts is good.

**Comment by jafingerhut** [comment]
> It also adds a PTF DPDK testing library which does not rely on p4runtime shell and instead uses Protobuf messages directly. @Hoooao implemented all of this.

Is there any difference between the Python code in this PR for creating/parsing Protobuf messages, and that used for any target, e.g. BMv2, other than DPDK?  It seems beneficial to have as few different Python libraries for this as we can get down to.

**Comment by fruffy** [comment]
> > It also adds a PTF DPDK testing library which does not rely on p4runtime shell and instead uses Protobuf messages directly. @Hoooao implemented all of this.
> 
> Is there any difference between the Python code in this PR for creating/parsing Protobuf messages, and that used for any target, e.g. BMv2, other than DPDK? It seems beneficial to have as few different Python libraries for this as we can get down to.

There are subtle differences in the two libraries, largely revolving around election id and device id. We could fix this in a separate PR and add a common P4Runtime communication library?

**Comment by fruffy** [comment]
> > It also adds a PTF DPDK testing library which does not rely on p4runtime shell and instead uses Protobuf messages directly. @Hoooao implemented all of this.
> 
> Is there any difference between the Python code in this PR for creating/parsing Protobuf messages, and that used for any target, e.g. BMv2, other than DPDK? It seems beneficial to have as few different Python libraries for this as we can get down to.

The changes were easy enough to do them here. We now have one single P4Runtime Python library that we can use for both BMv2 and DPDK. Will do some more cleanup for this PR. We will also have to fix the deprecated [add-on-miss](https://github.com/p4lang/p4c/blob/main/testdata/p4_16_samples/pna-dpdk-add_on_miss0.py) test.

File: backends/dpdk/run-dpdk-ptf-test.py [hunk 5]
```diff
@@ -72,15 +72,17 @@
     help="The log level to choose.",
 )
 
-# 9559 is the default P4Runtime API server port
-GRPC_PORT: int = 9559
+# 28000 is the default P4Runtime API server port
+GRPC_PORT: int = 28000
```
Comment (jafingerhut) [inline backends/dpdk/run-dpdk-ptf-test.py:76]: I have no personal issue if you want to use a non-default port number for the P4Runtime API connection, but this change to the comment is definitely incorrect.  9559 is the default port, as mentioned in the P4Runtime API spec.

**Review by jafingerhut — APPROVED** [review]
Except for my latest minor comment about a comment that was changed to now be factually incorrect, this looks good to me.

**Comment by fruffy** [comment]
> Except for my latest minor comment about a comment that was changed to now be factually incorrect, this looks good to me.

Fixed. I believe 28000 was the old port.  I have created an issue (#4204) for the remaining TODOs. This will be done in a new PR.

**Pull request merged**